Feature/in mem mongo setup

gs-hg opened the pull request:
#### What type of PR is this?

Adding ability to spin-up In-memory MongoDB we can use for development/testing.

gs-hg committed:
initial commit or in-memory mongo setup

gs-hg committed:
forgot welcome.pure changes

ghub-real commented on legend-engine-pure-ide-light/src/main/java/org/finos/legend/engine/ide/PureIDELight.java:
```diff
@@ -101,14 +101,14 @@ private static void enableEngineIntegration()
 
         System.setProperty("alloy.test.server.host", "127.0.0.1");
-        System.setProperty("alloy.test.server.port", "6060");
+        System.setProperty("alloy.test.server.port", "6300");
```
Thought your previous instructions were to make it 6060 for enabling local debug??

gs-hg replied:
Indeed, but since then they have changed the legend-engine server port to 6300 ( from 6000) so this makes it more inline with finos/master https://github.com/ghub-real/legend-engine/commit/0f37e3610c9dfcefed3faf82b05b2a15a2ea4348

ghub-real commented on legend-engine-xt-documentStore-executionPlan-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/nonrelational/MongoDbResource.java:
```diff
@@ -217,7 +254,7 @@ public Response insertCollectionItemAsJson(@QueryParam("collectionName") String
     private MongoDatabase getDefaultDB()
     {
-        return this.mongoDbClient.getMongoDBClient().getDatabase(DEFAULT_DATABASE_NAME);
+        return this.mongoDbClient.mongoClient.getDatabase(DEFAULT_DATABASE_NAME);
```
should we make this this.NonRelationalClient.getClient().getDatabase()
wondering with all Non-Relational naming do we want to make the objet MongoDbClient?   Ok either way, but would be good to think about it.

gs-hg replied:
I would keep this as MongoDbClient as that is clear and not tied down to anything else. ( can also see us having multiple clients in the future ie. DynamoDbClient ) Happy to have a NonRelationalClientProvider abstraction though

ghub-real replied:
There are two parts here:
this.mongoDbClient.getMongoDBClient()
my quesion is, should this.mongoDbClient - be a generic interface  & within that you have 
mongoClient which is specific.    else it is two layers for the same thing.     Am not sure we have the class hierarchy right here.  There is an interface + abstract class + 2 implementation.
Either we move the functions we use to the interface - so that Mongo DB becomes an instance object within the implementation.  Trying to c lean this up, so wecan pull to master

gs-hg replied:
NonRelationalClient is now the generic interface, abstract is simply so we don't duplicate logic of InMem vs ExternalMongoClients for nativeQueryExecution, happy to move those to both these implementations if we prefer

gs-hg committed:
code review comments, cleaning up class hierarchy

gs-hg committed:
review comments

ghub-real commented on legend-engine-xt-documentStore-executionPlan-connection/pom.xml:
```diff
@@ -71,6 +71,18 @@
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-databind</artifactId>
         </dependency>
+        <dependency>
```
version number?  do you want to specify in the root/pom.xml?  under dependencies??

gs-hg replied:
Good point, have added this, version already specified in pom.xml

ghub-real commented on legend-engine-xt-documentStore-executionPlan-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/nonrelational/AlloyMongoServer.java:
```diff
@@ -0,0 +1,56 @@
+        LOGGER.debug("Starting setup of connection for local Mongo database server on port: " + port + " and address: " + DEFAULT_MONGO_HOSTNAME);
+        MongoServer mongoServer = new MongoServer(new MemoryBackend());
+        mongoServer.bind(DEFAULT_MONGO_HOSTNAME, port);
+        System.out.println("Completed setup for local Mongo database server on port: " + port + " and address: " + DEFAULT_MONGO_HOSTNAME);
```
any specific reason for both logger & system.out?

gs-hg replied:
Missed this, using only LOGGER now

ghub-real commented on legend-engine-xt-documentStore-executionPlan-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/nonrelational/AlloyMongoServer.java:
```diff
@@ -0,0 +1,56 @@
+    protected static final String DEFAULT_MONGO_HOSTNAME = "localhost";
+    protected static final int DEFAULT_MONGO_PORT = 27017;
+
+    public static MongoServer startServer(int port)
```
do we need stopServer as well here? or something equivalent?

gs-hg replied:
have done so now.

ghub-real commented on legend-engine-xt-documentStore-executionPlan-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/nonrelational/client/NonRelationalClient.java:
```diff
@@ -24,4 +24,6 @@
 
     List<String> executeNativeQuery(String mongoQuery);
+
+    void shutDown();
```
should start Server also be here?? because it seems a bit inconsistent.  Worth checking what Connection interface providdes

gs-hg replied:
With Connection ,a Statement is created and that is what is used to execute SQL queries. (RelationalExecutor.java line 323 for example) On the relational side, HikariCP is used to create HikariDataSources and Hikari manages the connection pools. There is a ConnectionStateManagerPOJO concept but if we want to follow that it seems this is tied to the usage of HikariDataSource and DataSourceSpecification. 

To me, shutDown is needed for internal clean up prior to no longer needing the NonRelationalClient, if we initialise the client I don't see why we need an explicit initialise/begin/start() method, but we can add it. We will need to think about Connection pooling at some point but for our first cut here I wouldn't add too much complexity/management.

ghub-real replied:
the Mongo drivers do that, so we don't need any other library like hikari for that purpose.
mongo  client has a .close() wanted to check if we need to invoke it at all..  can come back to this later.

ghub-real reviewed (commented):
Looks fine - we can now easily carve out the mongoserver part into master

gs-hg committed:
PR comments:

- Added shut down for mongoserver ( in memory)

- Added root pom.xml versions to in-mem mongo dependencies

- Replaced usage of System.out.println -> sl4j LOGGER
